### PR TITLE
Protect against infinite macro expansion in def collector

### PR DESCRIPTION
**Pull request by flodiebold**
Something I noticed while trying to make macro expansion more resilient against errors.

There was a test for this, but it wasn't actually working because the first recursive expansion failed. (The comma...)

Even with this limit, that test (when fixed) still takes some time to pass because of the exponential growth of the expansions, so I disabled it and added a different one without growth.

CC @edwin0cheng

### CODE REVIEW

File: crates/ra_hir_def/src/nameres/collector.rs [hunk 1]
```diff
@@ -971,13 +988,26 @@ mod tests {
...
+    }
+
+    #[ignore] // this test does succeed, but takes quite a while :/
+    #[test]
+    fn test_macro_expand_will_stop_2() {
         do_resolve(
             r#"
         macro_rules! foo {
-            ($($ty:ty)*) => { foo!($($ty)*, $($ty)*); }
+            ($($ty:ty)*) => { foo!($($ty)* $($ty)*); }
```
Comment (edwin0cheng) [inline crates/ra_hir_def/src/nameres/collector.rs]: Oops, it should be `tt`, but I typed `ty` :(

File: crates/ra_hir_def/src/nameres/collector.rs [hunk 2]
```diff
@@ -552,8 +554,12 @@ where
         self.unexpanded_macros = macros;
         self.unexpanded_attribute_macros = attribute_macros;
 
-        for (module_id, macro_call_id) in resolved {
-            self.collect_macro_expansion(module_id, macro_call_id);
+        for (module_id, macro_call_id, depth) in resolved {
+            if depth > 128 {
```
Comment (edwin0cheng) [inline crates/ra_hir_def/src/nameres/collector.rs]: although the default rustc `recursion_limit is 128, but I think 1024 would be our [safe bet](https://grep.app/search?q=%23%21%5Brecursion_limit%20%3D%20&filter[lang][0]=Rust) here. 

OTOH, One day, we will set it via the `recursion_limit` of the  module `attr`.

**Comment by edwin0cheng** [comment]
Just a small nit, and otherwise LGTM.

File: crates/ra_hir_def/src/nameres/collector.rs (hunk 1 above)
Comment (flodiebold) [reply crates/ra_hir_def/src/nameres/collector.rs]: ah yeah, that would have worked as well

File: crates/ra_hir_def/src/nameres/collector.rs (hunk 2 above)
Comment (flodiebold) [reply crates/ra_hir_def/src/nameres/collector.rs]: I reduced it to try to get the second test to pass in an acceptable time, but it wasn't enough anyway.

Comment (edwin0cheng) [reply crates/ra_hir_def/src/nameres/collector.rs]: Yeah, even [playground](https://play.rust-lang.org/) refused to run too :)

**Pull request merged**